misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/BTestRunner.java

```diff
@@ -516,61 +515,110 @@ private void invokeDataDrivenTest(TestSuite suite, String testName, String key,
+                                               AtomicBoolean shouldSkipTest, List<String> failedOrSkippedTests,
+                                               List<String> failedAfterFuncTests) {
+        TesterinaResult functionResult;
+        // Check if the test depends on other failed test functions or after functions
+        if (isTestDependsOnFailedFunctions(test.getDependsOnTestFunctions(), failedOrSkippedTests) ||
+                isTestDependsOnFailedFunctions(test.getDependsOnTestFunctions(), failedAfterFuncTests)) {
+            shouldSkipTest.set(true);
+        }
+        // Check if test should be skipped
         if (!shouldSkip.get() && !shouldSkipTest.get()) {
-            Object valueSets = null;
-            if (test.getDataProvider() != null) {
-                valueSets = invokeTestFunction(suite, test.getDataProvider(), classLoader, scheduler);
-            }
-            if (valueSets == null) {
-                valueSets = invokeTestFunction(suite, test.getTestName(), classLoader, scheduler);
-                computeFunctionResult(test.getTestName(), packageName, shouldSkip, failedOrSkippedTests, valueSets);
-            } else {
-                if (valueSets instanceof BMap) {
-                    // Handle map data sets
-                    if (((BMap) valueSets).isEmpty()) {
-                        computeFunctionResult(test.getTestName(), packageName, shouldSkip, failedOrSkippedTests,
-                                new Error("The provided data set is empty."));
-                    } else {
-                        List<String> keyValues = getKeyValues((BMap) valueSets);
-                        Class<?>[] argTypes = extractArgumentTypes((BMap) valueSets);
-                        List<Object[]> argList = extractArguments((BMap) valueSets);
-                        for (int i = 0, argListSize = argList.size(); i < argListSize; i++) {
+            Object valueSets = invokeTestFunction(suite, test.getDataProvider(), classLoader, scheduler);
+            // Handle map datasets
+            if (valueSets instanceof BMap) {
+                // Handle map data sets
+                if (((BMap) valueSets).isEmpty()) {
+                    computeFunctionResult(test.getTestName(), packageName, shouldSkip, failedOrSkippedTests,
+                            new Error("The provided data set is empty."));
+                } else {
+                    List<String> keyValues = getKeyValues((BMap) valueSets);
+                    Class<?>[] argTypes = extractArgumentTypes((BMap) valueSets);
+                    List<Object[]> argList = extractArguments((BMap) valueSets);
+                    for (int i = 0, argListSize = argList.size(); i < argListSize; i++) {
+                        // Reset shouldSkipTest for each iteration, since Before function may not fail for other
+                        // iterations
+                        shouldSkipTest.set(false);
```

Comment:
Before function has the same logic right(It's not specific to the data that gets passed to the data-driven test) ? If it fails once, it's likely to fail in each iteration too.
I don't think we should reset.

Reply:
As discussed in the review we are special casing the data driven testcases so we need this particular reset to be implemented